### DISCUSSION TITLE
Make Malt Shading Essentials default plugin?

### DISCUSSION THREAD

**Discussion by Kolupsy** [bnpr/Malt#343] [Q&A]
I think it might be a good idea to have the shading essentials 'ship' with malt as an optional plugin, of course its more or less a project in of itself so maybe a submodule (have never worked with those so I am not entirely sure how that works) or maybe just aspects of it could be added to malt? I dont care about taking ownership of any of that stuff that I have made, I just care about people having a better time using and transitioning to malt. 
What do you think? Maybe a compromise where you can make decisions about the design of it (if you care to). Also does not have to be RIGHT NOW, but maybe when this commit: a165ae570f2a0a1f29d07c6b31ce891a6f1290b1 hits the release version or something

**Comment by dibli-goost** [comment]
I agree with this. At the very least, a plugin that ships with Malt could be useful as an example of how to make plugins for the engine?

I think it'll benefit Malt in the long run if it's a part of it, fed up EEVEE users are gonna be the main audience coming over, and having it be more 1:1 from the get go would draw them in, getting more users in early to test the engine and further spur developments.

**Comment by pragma37** [comment]
By default plugin, do you mean enabled by default or just ship it with Malt?

In any case, this would require setting up a default plugin directory for the user and a way to enable/disable individual plugins.
In the case of the Shading Essentials, it would be tricky to allow enabling/disabling it at runtime due to the crashes when registering/unregistering Blender nodes at runtime.

For enabling it by default, my main concern is that the Shading Essentials duplicate some functionality that is already available in Malt, and splits the node UI in two. So while I certainly like the advantages it provides to the user experience, having that kind of duplication by default feels kind of weird.

Honestly, I think it would be best to simply integrate it into Malt, improve the META system so the BlenderMalt side of things can be kept as minimal as possible, and expose all the built-in functionality in a more cohesive way.
You're doing a great job on the nodes and UX side of things, so if you like the idea, I think it would be best for Malt if you take the lead on that front.

**Reply by Kolupsy** [reply]
> By default plugin, do you mean enabled by default or just ship it with Malt?

I mean ship with malt by default

> In any case, this would require setting up a default plugin directory for the user and a way to enable/disable individual plugins

Malt already has a default plugins folder, right? (https://github.com/bnpr/Malt/tree/Development/plugins)

> In the case of the Shading Essentials, it would be tricky to allow enabling/disabling it at runtime due to the crashes when registering/unregistering Blender nodes at runtime.

That is true, its a valid concern. What I would propose in this case is to actually have it enabled at all time, just having an option to enabled/disable it in every place where it affects the UI (which I think currently is only the Add menu in the Malt Tree)

> Honestly, I think it would be best to simply integrate it into Malt

That would be great too! Although I am not sure how that would work. I can see all the shader code being integrated into Malt but what about all the python code that handles the node UI and more proprietary systems like the drivers manager? (https://github.com/Kolupsy/MaltShadingEssentials/blob/main/driver_registry.py)
I think a design change in malt could replace this particular example but in the future i might come up with other systems that have a limited scope and would make more sense as a plugin. 

> I think it would be best for Malt if you take the lead on that front.

Absolutely would love to. I do think if we could discuss the things I mentioned above that would give me a clearer idea of how to actually approach it.

**Reply by pragma37** [reply]
Nice, maybe we could set up a call to discuss it?

> Malt already has a default plugins folder, right? 

But this is not even shipped with the addon releases.

**Reply by Kolupsy** [reply]
> But this is not even shipped with the addon releases.

Oooh haha I hadnt even realized. Interesting, so 'default plugins' dont even exist then. That definitely changes up my original plans no matter what.

> Nice, maybe we could set up a call to discuss it?

Sounds like a good idea...

**Comment by HSDSZ** [comment]
It seems that https://github.com/bnpr/Malt/pull/348 is stable and ready for testing?

**Reply by pragma37** [reply]
It's ready for testing and I plan to ask for user feedback ASAP.
But it's not stable. There can be bugs and there could be breaking changes.
